### PR TITLE
Padaquad integration

### PR DESCRIPTION
**Pull request by khegazy**

### CODE REVIEW

File: popcornn/tools/integrator.py [hunk 1]
```diff
@@ -219,87 +329,104 @@ def integrate_path(
...
+            ``[N, 1]`` (already 2-D) — no unsqueeze. ``dE/dt = -(F·v).sum(-1)``
+            reuses the forces/velocities already resolved for the loss."""
             l, variables = evaluate_integrand_sum(
-                self._terms,
-                t_flat.unsqueeze(-1),
-                path,
-                also_resolve=also_resolve,
+                self._terms, t, path, also_resolve=also_resolve,
             )
+            l_per_t = l.reshape(t.shape[0], -1).sum(dim=-1)  # [N], graph live
```
Comment (khegazy) [inline popcornn/tools/integrator.py]: Print out shapes and check that we are not summing numbers, but instead removing a dimension.

File: popcornn/tools/integrator.py [hunk 2]
```diff
@@ -219,87 +329,104 @@ def integrate_path(
...
+                return jac, tracked
+            take_gradient = False
+        else:
+            # Gradient-of-the-integral: the integrand returns the scalar loss
+            # [N, 1] with the graph live; padaquad integrates it with
+            # take_gradient=True and calls .backward() on the loss integral
+            # per batch, accumulating ``∫∂L/∂θ dt`` into param.grad via autograd.
+            def f(t):
+                l_per_t, tracked = _loss_and_tracked(t)
+                return l_per_t.reshape(-1, 1), tracked  # [N, 1], graph live
```
Comment (khegazy) [inline popcornn/tools/integrator.py]: Check this shape as well.

File: popcornn/tools/integrator.py [hunk 3]
```diff
@@ -219,87 +329,104 @@ def integrate_path(
...
+            take_gradient = True
+
+        # Whether padaquad must benchmark the integrand's memory this call.
+        # The closure ``f`` is rebuilt every call, so padaquad's id(f)-keyed
+        # cache never hits; passing an explicit ``max_batch`` suppresses the
+        # benchmark. On the first call (max_batch is None) we let it run, then
+        # snapshot the sizing so subsequent calls skip it — benchmarking once
+        # per integrator lifetime rather than once per optimizer step.
+        snapshot_max_batch = self.max_batch is None
```
Comment (khegazy) [inline popcornn/tools/integrator.py]: This should be gone.

File: popcornn/tools/integrator.py [hunk 4]
```diff
@@ -219,87 +329,104 @@ def integrate_path(
...
-            t_final_0d,
-            method=self.method,
-            atol=self.atol,
-            rtol=self.rtol,
-            norm=self.norm,
+            mesh=self._mesh_optimal,          # None on first call -> fresh random mesh
+            mesh_init=t_init_1d,
+            mesh_final=t_final_1d,
+            error_norm=self.norm,
+            error_integral_reference=self._previous_integral,
```
Comment (khegazy) [inline popcornn/tools/integrator.py]: delete

File: popcornn/tools/integrator.py [hunk 5]
```diff
@@ -219,87 +329,104 @@ def integrate_path(
...
-            offset += k
-
-        # No scalar loss graph in this design. Surface ‖∫∇L dt‖_∞ (per-component
-        # max) as the convergence signal consumed by PathOptimizer's threshold
-        # check. L∞ is closer to MLP-size-independent than L2 — the latter scales
-        # with √D and forces the threshold to be retuned per parameter count.
-        # Also expose L2 for monitoring (cheap on the same flat tensor).
+        # Warm-start the next call from this run's pruned + refined mesh.
+        self._mesh_optimal = result.mesh_optimal
+        self._previous_integral = result.integral.detach()
```
Comment (khegazy) [inline popcornn/tools/integrator.py]: delete

File: popcornn/tools/integrator.py [hunk 6]
```diff
@@ -219,87 +329,104 @@ def integrate_path(
...
+        else:
+            # padaquad already called .backward(), accumulating ∫∂L/∂θ dt into
+            # param.grad. Gather it into a flat [D] vector for grad_norm and the
+            # grad_integral alias. (Frozen/untouched params contribute zeros.)
+            flat = torch.cat([
+                p.grad.reshape(-1) if p.grad is not None
+                else p.new_zeros(k).reshape(-1)
+                for p, k in zip(params, sizes)
+            ]).detach()
+        integral_output.grad_integral = flat
```
Comment (khegazy) [inline popcornn/tools/integrator.py:428]: Take a look at dataclass and the difference between integral and grad_integral.

File: popcornn/tools/integrator.py [hunk 7]
```diff
@@ -308,59 +435,61 @@ def f(t_flat):
...
-                rtol=self.loss_rtol,
-                norm=self.norm,
-                max_batch=self.max_batch,
-                device=self.device,
-                dtype=self.dtype,
+                mesh=self._loss_mesh_optimal,
+                mesh_init=t_init_1d,
+                mesh_final=t_final_1d,
+                error_norm=self.norm,
+                error_integral_reference=self._loss_previous_integral,
```
Comment (khegazy) [inline popcornn/tools/integrator.py]: delete

File: popcornn/tools/integrator.py [hunk 8]
```diff
@@ -308,59 +435,61 @@ def f(t_flat):
...
+                take_gradient=False,
+                max_batch=self.max_batch,     # int by now (snapshotted on the grad pass)
             )
-            integral_output.loss = loss_output.integral.detach()
-            # Loss pass may shrink further (it has no autograd graph, but
-            # the integrand magnitudes can drive different K's per interval).
-            self.max_batch = loss_output.max_batch
+            self._loss_mesh_optimal = loss_result.mesh_optimal
+            self._loss_previous_integral = loss_result.integral.detach()
```
Comment (khegazy) [inline popcornn/tools/integrator.py]: delete

File: popcornn/tools/integrator.py (hunk 1 above)
Comment (khegazy) [reply popcornn/tools/integrator.py]: Summation is for taking the gradient; it needs to be a scalar. Moved this to the gradient call.

File: popcornn/tools/integrator.py (hunk 2 above)
Comment (khegazy) [reply popcornn/tools/integrator.py]: fixed as well.

File: popcornn/tools/integrator.py (hunk 3 above)
Comment (khegazy) [reply popcornn/tools/integrator.py]: removed

File: popcornn/tools/integrator.py (hunk 6 above)
Comment (khegazy) [reply popcornn/tools/integrator.py:428]: This is flattening the gradients to be used later. Shifts the output of take_gradient=True/False into the same form.
